holmes/plugins/toolsets/azure_sql/azure_sql_toolset.py

```diff
@@ -0,0 +1,182 @@
+                GetTopCPUQueries(self),
+                GetSlowQueries(self),
+                GetTopDataIOQueries(self),
+                GetTopLogIOQueries(self),
+                GetActiveAlerts(self),
+                AnalyzeConnectionFailures(self),
+            ],
+        )
+        self._reload_llm_instructions()
+
+    def prerequisites_callable(self, config: Dict[str, Any]) -> Tuple[bool, str]:
+        if not config:
+            return False, TOOLSET_CONFIG_MISSING_ERROR
+
+        errors = []
+        try:
+            azure_sql_config = AzureSQLConfig(**config)
+
+            # Set up Azure credentials
+            try:
+                credential: Union[ClientSecretCredential, DefaultAzureCredential]
+                if (
+                    azure_sql_config.tenant_id
+                    and azure_sql_config.client_id
+                    and azure_sql_config.client_secret
+                ):
+                    logging.info(
+                        "Using ClientSecretCredential for Azure authentication"
+                    )
+                    credential = ClientSecretCredential(
+                        tenant_id=azure_sql_config.tenant_id,
+                        client_id=azure_sql_config.client_id,
+                        client_secret=azure_sql_config.client_secret,
+                    )
+                else:
+                    logging.info(
+                        "Using DefaultAzureCredential for Azure authentication"
+                    )
+                    credential = DefaultAzureCredential()
+
+                # Test the credential by attempting to get tokens for both required scopes
+                mgmt_token = credential.get_token(
+                    "https://management.azure.com/.default"
+                )
+                if not mgmt_token.token:
+                    raise Exception("Failed to obtain Azure management token")
+
+                # Test SQL database token as well
+                sql_token = credential.get_token(
+                    "https://database.windows.net/.default"
+                )
+                if not sql_token.token:
+                    raise Exception("Failed to obtain Azure SQL database token")
+
+            except Exception as e:
+                message = f"Failed to set up Azure authentication: {str(e)}"
+                logging.error(message)
+                errors.append(message)
+                return False, message
+
+            # Store single database configuration and create API client
+            self._database_config = azure_sql_config.database
+            self._api_client = AzureSQLAPIClient(
+                credential, azure_sql_config.database.subscription_id
+            )
+            logging.info(
+                f"Configured Azure SQL database: {azure_sql_config.database.server_name}/{azure_sql_config.database.database_name}"
+            )
+
+            # Validate each tool's configuration requirements
```

Comment:
Just to make sure this is on purpose, we don't validate the tools then? we will add it in the future?

Reply:
This is actually a bigger problem.
I think we should validate the tools because each tool may require different permissions.
It makes sense that these tools are part of the same toolset. However toolsets are reported as fully working or not working at all in the SaaS UI. We need a third state which is `error reported but toolset enabled`.

This would allow reporting incorrect permissions, incorrect config for some tools without impacting other tools in the same toolset.

So for now we don't want to validate that all tools are working because if any permission/setup is incorrect it would prevent the whole toolset from working.